File: fft/go/fft/cooley_tukey.go
package fft

import "math"

func FFT(arr []Complex) {
	fft(arr)
	factor := 1.0 / math.Sqrt(float64(len(arr)))
	for i := range len(arr) {
		arr[i] = arr[i].MulScalar(factor)
	}
}

func fft(arr []Complex) {
	n := len(arr)
	if n == 1 {
		return
	}

	a0 := make([]Complex, 0, n/2)
	a1 := make([]Complex, 0, n/2)
	for i := range n / 2 {
		a0 = append(a0, arr[2*i])
		a1 = append(a1, arr[2*i+1])
	}

	fft(a0)
	fft(a1)

	ang := -2 * math.Pi / float64(n)
	w := Complex{1, 0}
	wn := Complex{math.Cos(ang), math.Sin(ang)}
	for i := range n / 2 {
		p := a0[i]
		q := w.Mul(a1[i])
		arr[i] = p.Add(q)
		arr[i+n/2] = p.Sub(q)
		w = w.Mul(wn)
	}
}

File: fft/swift/Sources/FFT/CooleyTukey.swift
import Foundation

public func fft(_ arr: inout [Complex]) {
  func _fft(_ arr: inout [Complex]) {
    let n = arr.count
    if n == 1 {
      return
    }

    var a0 = [Complex]()
    a0.reserveCapacity(n / 2)
    var a1 = [Complex]()
    a1.reserveCapacity(n / 2)
    for i in 0..<n / 2 {
      a0.append(arr[2 * i])
      a1.append(arr[2 * i + 1])
    }

    _fft(&a0)
    _fft(&a1)

    let ang = -2.0 * Double.pi / Double(n)
    var w = Complex(real: 1, imag: 0)
    let wn = Complex(real: cos(ang), imag: sin(ang))
    for i in 0..<n / 2 {
      let p = a0[i]
      let q = w * a1[i]
      arr[i] = p + q
      arr[i + n / 2] = p - q
      w = w * wn
    }
  }

  _fft(&arr)
  let factor = 1.0 / Double(arr.count).squareRoot()
  for i in 0..<arr.count {
    arr[i] = arr[i] * factor
  }
}
